cEP-0021.md

```diff
@@ -0,0 +1,133 @@
+
+## Abstract
+
+This cEP describe the detailed implementation of profiling interface as a part
+of the
+[GSoC'18 project](https://summerofcode.withgoogle.com/projects/#6109762077327360).
+
+## Introduction
+
+Profiling a Python program is a form of dynamic analysis that measures the
+execution time of the program and the functions it is composed of. It's goal is
+to produce data about where our program is spending time, and what area might
+be worth optimizing. Thus we can provide a *Profiling interface* from within
+coala and make it available to bear writers so that they can create better
+performant bears.
+
+This cEP proposes to introduce the profiling interface so that bear writers
+will have the ability to Profile the Bear's code to optimize it's
+performance.
+
+## Proposed Change
+
+Bear writers can invoke the profiler with the `--profile-bears` argument.
+
+Bear writer will have the ability to directly dump the raw
+profile output either on *current working <br> directory* or to a
+*specified directory name*. (which can be further used for estimation
+of profiler stats)
+
+- Using default settings provided by `coala` i.e
+  `coala --profile-bears` or `profile_bears = True`
+  (it will dump the raw profile output in *current working directory*).
+- By passing the directory name through *CLI* i.e. <br>
+  `coala --profile-bears [Dirname]` or using <br>
+  `.coafile` i.e. `profile_bears = [Dirname]`
+  It will dump the *raw profile output* on the specified
+  directory, i.e., [Dirname]
+
+The profiling will start by profiling the `run()` method of the bears
+The advantage of profiling just the `run()` method is that, bear writer
+doesn't have to spend time on loading the files, collecting the settings,
+etc. as this is all done by coala itself, they just have to Profile
+the `run()` method to create a Better Performant bear.
+
+# Implementation Details
+
+When user invokes the profiler with either `CLI` or by `coafile`
+then `execute` method of Bear base Class will call the `run_bear_from_section`
+method.
+
+```python
+def run_bear_from_section(self, args, kwargs):
+    ............
+    profile_bears_results, args, kwargs = self.setup_profiler(
+        *args, **kwargs)
+    return profile_bears_results if profile_bears_results else self.run(
+        *args, **kwargs)
+```
+
+Which further calls the `setup_profiler()` method, and `setup_profiler()`
+method check whether keyword arguments passed to the Bear contain
+the `profile_bears` key or not.
+
+```python
+if profile_bears is None:
+    kwargs.pop('profile_bear')
+    return False, args, kwargs
+```
+
+- If `yes` then it will check the value of `profile_bears` key is
+  `True` or `[Dirname]`
+  - If `True` then it will create a profiler file for every bear of
+    every section with a name `{section_name}{bear_name}.prof` and dump
+    it in a *current working directory.*
+
+```python
+if str(profile_dump) is u'False' :
```

Comment:
We are working with python 3, you don't need `u` ;)
Also you should compare using `==`

But can be fixed after merge

Reply:
Forgot to write `not` here.